wazo_dird/plugins/office365_backend/plugin.py

```diff
@@ -0,0 +1,161 @@
+            logger.info(
+                'no "searched_columns" configured on "%s" no results will be matched',
+                self.name,
+            )
+
+        self._first_matched_columns = config.get(self.FIRST_MATCHED_COLUMNS, [])
+        if not self._first_matched_columns:
+            logger.info(
+                'no "first_matched_columns" configured on "%s" no results will be matched',
+                self.name,
+            )
+
+    def search(self, term, args=None):
+        logger.debug('Searching term=%s', term)
+        try:
+            microsoft_token = self._get_microsoft_token(**args)
+        except MicrosoftTokenNotFoundException:
+            return []
+
+        contacts = self.office365.get_contacts_with_term(microsoft_token, term, self.endpoint)
+        updated_contacts = self._update_contact_fields(contacts)
+
+        lowered_term = term.lower()
+
+        def match_fn(contact):
+            for column in self._searched_columns:
+                column_value = contact.get(column) or ''
+                if lowered_term in str(column_value).lower():
+                    return True
+            return False
+
+        filtered_contacts = [c for c in updated_contacts if match_fn(c)]
+        sorted_contacts = sorted(filtered_contacts, key=itemgetter('givenName'))
+
+        return [self._SourceResult(c) for c in sorted_contacts]
+
+    def list(self, unique_ids, args=None):
+        try:
+            microsoft_token = self._get_microsoft_token(**args)
+        except MicrosoftTokenNotFoundException:
+            return []
+
+        contacts = self.office365.get_contacts(microsoft_token, self.endpoint)
+        updated_contacts = self._update_contact_fields(contacts)
+        filtered_contacts = [c for c in updated_contacts if c[self.unique_column] in unique_ids]
+
+        return [self._SourceResult(contact) for contact in filtered_contacts]
+
+    def first_match(self, term, args=None):
+        if not self._first_matched_columns:
+            logger.debug(
+                '%s is a source for reverse lookups but the not have a "first_matched_columns"',
+                self.name,
+            )
+            return
+
+        try:
+            microsoft_token = self._get_microsoft_token(**args)
+        except MicrosoftTokenNotFoundException:
+            logger.debug('could not find a matching microsoft token, aborting first_match')
+            return None
+
+        contacts = self.office365.get_contacts(microsoft_token, self.endpoint)
+        updated_contacts = self._update_contact_fields(contacts)
+        lowered_term = term.lower()
+
+        for contact in updated_contacts:
+            if self._first_match_predicate(lowered_term, contact):
+                return self._SourceResult(contact)
+
+    def _first_match_predicate(self, term, contact):
+        for column in self._first_matched_columns:
+            column_value = contact.get(column) or ''
+
+            if not isinstance(column_value, list):
```

Comment:
Starting with the True expression first is more intuitive, e.g:

`if isinstance(column_value, list):`

Reply:
I agree